src/boot/efi/stub.c

```diff
@@ -632,13 +697,34 @@ static EFI_STATUS run(EFI_HANDLE image) {
                 pcrpkey_initrd = mfree(pcrpkey_initrd);
         }
 
+        /* First load the base device tree, then fix it up using addons - global first, then per-UKI. */
         if (dt_size > 0) {
                 err = devicetree_install_from_memory(
                                 &dt_state, PHYSICAL_ADDRESS_TO_POINTER(dt_base), dt_size);
                 if (err != EFI_SUCCESS)
                         log_error_status(err, "Error loading embedded devicetree: %m");
         }
 
+        for (size_t i = 0; i < n_dts_addons_global; ++i) {
+                err = devicetree_install_from_memory(
+                                &dt_state,
+                                dt_bases_addons_global[i],
+                                dt_sizes_addons_global[i]);
+                if (err != EFI_SUCCESS)
+                        log_error_status(err, "Error loading global addon devicetree: %m");
+                dt_bases_addons_global[i] = mfree(dt_bases_addons_global[i]);
+        }
+
+        for (size_t i = 0; i < n_dts_addons_uki; ++i) {
+                err = devicetree_install_from_memory(
+                                &dt_state,
+                                dt_bases_addons_uki[i],
+                                dt_sizes_addons_uki[i]);
+                if (err != EFI_SUCCESS)
+                        log_error_status(err, "Error loading local addon devicetree: %m");
+                dt_bases_addons_uki[i] = mfree(dt_bases_addons_uki[i]);
+        }
```

Comment:
maybe add a helper: install_dt_addons() that takes the array and the number of entries as parameters. i.e. unify these two almost identical for loops in a single function